core/src/main/java/org/apache/iceberg/InternalData.java

```diff
@@ -0,0 +1,169 @@
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class InternalData {
+  private InternalData() {}
+
+  private static final Logger LOG = LoggerFactory.getLogger(InternalData.class);
+  private static final Map<FileFormat, Function<OutputFile, WriteBuilder>> WRITE_BUILDERS =
+      Maps.newConcurrentMap();
+  private static final Map<FileFormat, Function<InputFile, ReadBuilder>> READ_BUILDERS =
+      Maps.newConcurrentMap();
+
+  static void register(
+      FileFormat format,
+      Function<OutputFile, WriteBuilder> writeBuilder,
+      Function<InputFile, ReadBuilder> readBuilder) {
+    WRITE_BUILDERS.put(format, writeBuilder);
+    READ_BUILDERS.put(format, readBuilder);
+  }
+
+  @SuppressWarnings("CatchBlockLogException")
+  private static void registerSupportedFormats() {
+    InternalData.register(
+        FileFormat.AVRO,
+        outputFile -> Avro.write(outputFile).createWriterFunc(InternalWriter::create),
+        inputFile -> Avro.read(inputFile).createResolvingReader(InternalReader::create));
+
+    try {
+      DynMethods.StaticMethod registerParquet =
+          DynMethods.builder("register")
+              .impl("org.apache.iceberg.InternalParquet")
+              .buildStaticChecked();
+
+      registerParquet.invoke();
+
+    } catch (NoSuchMethodException e) {
+      // failing to load Parquet is normal and does not require a stack trace
+      LOG.info("Unable to register Parquet for metadata files: {}", e.getMessage());
+    }
+  }
+
+  static {
+    registerSupportedFormats();
+  }
+
+  public static WriteBuilder write(FileFormat format, OutputFile file) {
+    Function<OutputFile, WriteBuilder> writeBuilder = WRITE_BUILDERS.get(format);
+    if (writeBuilder != null) {
+      return writeBuilder.apply(file);
+    }
+
+    throw new UnsupportedOperationException(
+        "Cannot write using unregistered internal data format: " + format);
+  }
+
+  public static ReadBuilder read(FileFormat format, InputFile file) {
+    Function<InputFile, ReadBuilder> readBuilder = READ_BUILDERS.get(format);
+    if (readBuilder != null) {
+      return readBuilder.apply(file);
+    }
+
+    throw new UnsupportedOperationException(
+        "Cannot read using unregistered internal data format: " + format);
+  }
+
+  public interface WriteBuilder {
+    /** Set the file schema. */
+    WriteBuilder schema(Schema schema);
+
+    /** Set the file schema's root name. */
+    WriteBuilder named(String name);
+
+    /**
+     * Set a writer configuration property.
+     *
+     * <p>Write configuration affects writer behavior. To add file metadata properties, use {@link
+     * #meta(String, String)}.
+     *
+     * @param property a writer config property name
+     * @param value config value
+     * @return this for method chaining
+     */
+    WriteBuilder set(String property, String value);
```

Comment:
Is it intentional that we have `meta(Map)`, but we don't have `set(Map)`?

Reply:
`meta(Map)` is used in the code paths that call this. We can also plumb through `setAll` but I was going for a minimal set of methods here.